File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
from collections import Counter

def solution(want, number, discount):
    # 딕셔너리 초기화 key:want, value:number
    want_dic = {key:value for key, value in zip(want, number)}
    possible_date = 0
        
    # 없으면 배열 다음 인덱스로
    for i in range(10, len(discount) + 1): # len(want) 만큼 배열 잘라서 순회
        sale_list = discount[i - 10:i]
        
        # if set(sale_list) == set(want): # sale_list에 want 목록이 다 있는지 확인
        if want_dic == dict(Counter(sale_list)): # 있으면 물건 개수 체크
            possible_date += 1
    
    return possible_date

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        Map<String, Integer> wants = new HashMap<>();
        Map<String, Integer> glocery = new HashMap<>();
        
        int answer = 0;
        int n = discount.length;
        
        for (int i = 0; i < want.length; i++) wants.put(want[i], number[i]);
        for (int i = 0; i < 10; i ++) {
            glocery.put(discount[i], glocery.getOrDefault(discount[i], 0) + 1);
        }
        
        if (isValid(want, wants, glocery)) answer++;
        
        for (int start = 1; start <= n - 10; start++) {
            glocery.put(discount[start - 1], glocery.getOrDefault(discount[start - 1], 0) - 1);
            glocery.put(discount[start + 9], glocery.getOrDefault(discount[start + 9], 0) + 1);
            
            if (isValid(want, wants, glocery)) answer++;
        }
        
        
        return answer;
    }
    
    private boolean isValid(String[] want, Map<String, Integer> wants, Map<String, Integer> glocery) {
        for (int i = 0; i < want.length; i++) {
            String item = want[i];
            int need = wants.get(item);
            int have = glocery.getOrDefault(item, 0);
            
            if (have < need) return false;
        }
        
        return true;
    }
}